Add gvproxy to windows packages

arixmkii opened the pull request:
This adds gvproxy.exe (which is a requirement to use podman with Qemu on windows(https://github.com/containers/podman/issues/13006)) to Windows installers and zip builds.

Included changes:
* ~commit hash reference of gvisor-tap-vsock is bumped to latest (to include unix domain socket support on Windows);~ (was fixed already in upstream)
* wih-sshproxy build step is renamed to gvproxy (to match the repository name);
* gvproxy build step is responsible for both artifacts - win-sshproxy.exe and gvproxy.exe;
* gvproxy.exe added to zip archive;
* gvproxy.exe added to msi build;
* gvproxy.exe added to new Windows installer (unfortunately there is no fallback options as with win-sshproxy.exe, so, for packaging older versions - older version of script will be required);
* fixes for building on Windows with msys2:
  * GOROOT is passed to go executable for host detection (as msys packages cine with GOROOT override), this assumed to be safe for other builds;
  * ~`utf-8` encoding is enforced in python script;~ (was fixed already in upstream)
  * unix style line endings are enforced in python script;
  * added workaround with unlinking file before renaming in python script (as renaming behaves differently on Windows, when file exists);
  * removed explicit `-in /dev/stdin` from go-md2man as this a default (reads StdIn if not specified) and this triggers some regression in msys2 with latest updates.

This is extracted from my work on Qemu Podman Machine on Windows. It might look that there is no value in adding binary in advance, but I could challenge this with the statement, that it will be possible to launch gvproxy and Qemu VM manually and then setup podman remote connection to use it in a machine like scenario (without handy helpers of `podman machine`). If this can't be accepted int his form, I can move it to draft and add the description of this scenario to https://github.com/containers/podman/tree/main/docs/tutorials

Signed-off-by: Arthur Sengileyev <arthur.sengileyev@gmail.com>

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
gvproxy.exe added to Windows installers
```

arixmkii commented:
Added documentation how to setup QEMU machine with gvproxy and use it with podman.

arixmkii commented:
Squashed commits. Added one more change - removed explicit "-in /dev/stdin" from go-md2man because there is a regression in how msys2 handles this input after some updates.
```
 open /proc/self/fd/0: The system cannot find the path specified.
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+## Introduction
+
+There are officially supported podman machines with QEMU on macOS and WSL2 on Windows, but if one wants to use manually configured QEMU VM in a remote mode this could guide them through.
```
```suggestion
There are officially supported podman machines with QEMU on macOS and WSL2 on Windows, but if one wants to use a manually configured QEMU VM in remote mode, this guide should be helpful.
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+There are officially supported podman machines with QEMU on macOS and WSL2 on Windows, but if one wants to use manually configured QEMU VM in a remote mode this could guide them through.
+
+This writing will focus on Windows, because there is no official QEMU support provided, but it should be relatively easy to adapt them for macOS use.
```
```suggestion
This writing will focus on Windows because no official QEMU support is provided, but it should be relatively easy to adapt the following for macOS use if desired.
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+## Prerequisites
+
+* Windows 11 or 10 with unix domain sockets support
```
```suggestion
* Windows 11 or 10 with Unix domain sockets support
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+* SSH client feature installed on the machine
+* Hyper-V acceleration should be operational on the machine
+* Direcroty `C:\qemu-remote\` will be used for storing needed assets
+* Port `57561` free to use it for ssh over loopback interface
```
```suggestion
* Port `57561` is free to use for ssh over a loopback interface
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+Download QEMU (7.2.0 minimal) from https://qemu.weilnetz.de/w64/
+
+Then download the FCOS image for QEMU from https://getfedora.org/coreos/download?tab=metal_virtualized&stream=testing&arch=x86_64
```
Unless I missed that you spelled this out earlier.
```suggestion
Then download the Fedora CoreOS (FCOS) image for QEMU from https://getfedora.org/coreos/download?tab=metal_virtualized&stream=testing&arch=x86_64
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+One will need `.xz` format extraction tool like xz itself or 7-zip. Use it to extract the `.qcow2` image to C:\qemu-remote\fedora-coreos-37.20221127.2.0-qemu.x86_64.qcow2
+
+With xz the command line (when run from the same directory) will be `xz -d fedora-coreos-37.20221127.2.0-qemu.x86_64.qcow2.xz`
```
suggestion for consideration
```suggestion
With xz the command line (when run from the same directory) will be 
```
xz -d fedora-coreos-37.20221127.2.0-qemu.x86_64.qcow2.xz
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+### Podman
+
+Download or prepare Podman Windows build with `gvproxy` included and install as default.
```
Can you add links here for both of these options please?

arixmkii replied:
I reworked this section as build instructions for the Podman installer are extremely complex. So, I referenced the Podman releases and provided instructions on how to build `gvproxy.exe` if it is missing in the installer (releases before this PR is merged).

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+### SSH
+
+Generate ssh keys with empty passphrase
```
```suggestion
Generate ssh keys with an empty passphrase
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+### Ingition for FCOS
+
+Create ignintion file C:\qemu-remote\remote.ign with content of
```
```suggestion
Create ignition file C:\qemu-remote\remote.ign with the content of
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+Replace "YOURSSHKEYHERE" with the actual pub keys you generated.
+
+## Laucning
```
```suggestion
## Launching
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+### gvproxy
+
+One needs to run gvproxy first to make it ready for the QEMU VM launched afterwards. Run it  with the command below
```
Or a period at the end if you don't like the semi.
```suggestion
One needs to run gvproxy first to make it ready for the QEMU VM launched afterward. Run it  with the command below:
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+### QEMU
+
+Laucnh QEMU with the following command (it is configure for 4 CPUs and 8 GB RAM, but could be adjusted for less):
```
```suggestion
Launch QEMU with the following command (the following configures it to use 4 CPUs and 8 GB RAM, but it could be adjusted for less):
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+### First time launch extras
+
+Observe QEMU loading and wait for the message of SSH keys being provisioned to the machine. Next before making first ssh connection one would need to add it to known hosts.
```
```suggestion
Observe QEMU loading and wait for the message of SSH keys being provisioned to the machine. Next, before making the first ssh connection, one would need to add it to known hosts.
```

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
````diff
@@ -0,0 +1,123 @@
+ssh-keyscan -p 57561 127.0.0.1 >> %USERPROFILE%\.ssh\known_hosts
+```
+
+### Add new connection to podman
````
```suggestion
### Add new connection to Podman
```

TomSweeneyRedHat replied:
Except for actual commands, it should be "Podman" through out.

TomSweeneyRedHat commented on docs/tutorials/qemu-remote-tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+## Shutting down the machine
+
+Podman machine built in machinery will not work for custom machine. One would need to gracefully shutdown it connecting via SSH.
```
```suggestion
 The built-in machinery of  Podman machine will not work for a custom machine. One needs to gracefully shut it down by connecting via SSH:
```

arixmkii commented:
It is not forgotten. Will update the PR this week.

arixmkii commented:
Rebased. Description updates (as some fixes were already applied upstream).

@TomSweeneyRedHat I addressed the issues you pointed out in a separate commit. Taking into account amount feedback you contributed, should I mention you in `Co-authored-by:` section for that commit (I could not just accept suggestions, because I needed to rebase the code first)?

arixmkii commented:
gvproxy team promised to upload latest artifacts to the releases page. I will then update the section removing gvproxy build instructions and replace it with a link.

arixmkii commented:
@n1hility could you take a look at my changes for Windows installer? You are the only contributor for that part, probably, you are the best person to check my changes then. Thank you in advance!

arixmkii commented:
I'm leaning towards adding [NO NEW TESTS NEEDED] if there no tests, which validates the content of Zip or exe artifacts.

rhatdan commented:
SGTM, add no new test needed.

n1hility commented on Makefile:
```diff
@@ -757,16 +757,17 @@ podman-v%.msi: test/version/version
-.PHONY: win-sshproxy
-win-sshproxy: test/version/version
+.PHONY: gvproxy
+gvproxy: test/version/version
```
Since this is Windows specific, to avoid confusion, we probably should rename the Makefile target to win-gvproxy

n1hility reviewed (changes requested):
Looks good but have some suggestions here, mostly around docs. We were likely to include gvproxy.exe anyway for a future user-space network so this is a bit serendipitous!

arixmkii commented:
Updated tutorial, rebased to latest main and squashed commits.

Changes:
* applied recommended changes in tutorial;
* reworked and simplified Introduction section;
* updated Makefile.

@n1hility kept the changes since latest version as a separate commit in a different branch https://github.com/arixmkii/podman/commit/f785459772a5766cdce9a5e934156255cce8c438 for your convenience doing the review.

n1hility commented on contrib/win-installer/podman.wxs:
```diff
@@ -29,6 +29,9 @@
             <Component Id="WinSshProxyExecutable" Guid="0DA730AB-2F97-40E8-A8FC-356E88EAA4D2" Win64="yes">
               <File Id="WinSshProxyExecutableFile" Name="win-sshproxy.exe" Source="artifacts/win-sshproxy.exe" KeyPath="yes"/>
+            </Component>
+             <Component Id="GvProxyExecutable" Guid="1A4A2975-AD2D-44AA-974B-9B343C098333" Win64="yes">
```
@arixmkii just had a thought here, what if we just surrounded these gvproxy sections with a pre-processor conditional directive e.g. `<?if $(env.UseGVProxy) != Skip?>`

The powershell release process script can then just set it as a return value passed as a param to the build. If you don't have the bandwidth I could throw that on. Let me know.

arixmkii replied:
Added in a separate commit.

Implementation:
* In build script: check if there is no gvproxy, then don't try to sign it and set env var to Skip for child processes
* in Wix xml: conditionally check for presence of env var and if it exists then use it as a value for the variable, if not then just fallback to default value - empty string. This is needed to workaround undefined variable expansion, because to me it looked like missing env var should not fail the build.

arixmkii replied:
I didn't apply similar fix to old `podman.msi`, because building from the actual source tree it will have gvproxy.exe starting from the same commit.

n1hility replied:
Excellent. This change looks good to me. Great work. Just to confirm you manually tested both scenarios? 

I agree with the old msi not being necessary for this change, since its only built as a one-off. The plan is to remove it soon anyway.  It was just left behind until we knew the new installer was in wide use.

arixmkii replied:
Yes. I used zip archive built locally and the artifact from v4.3.1 stable to test both scenarios. And then installed it on my test system to verify extracted content in Program Files folder.
It actually took me a number of iterations to get both of them working 😅  So, here I'm really happy that I did these tests.

n1hility replied:
cool thanks. yeah it's amazing how finicky the installer technology stack is for Windows!

n1hility reviewed (approved):
/approve LGTM

vrothberg reviewed (commented):
/lgtm

rhatdan commented:
/cherry-pick v4.4
